### PR TITLE
Edit me link improvements

### PR DESCRIPTION
**Pull request by nik9000**
Makes two changes to the generation of "edit" links for asciidoctor:
1. Prefers the longest matching repository. This helps when repository
is `logtstash` and another is `logstash-docs`.
2. Create a setting to enable overriding `edit_me` links using an
attribute. Historically we've overridden this attribute in cases where
it is no longer needed for Asciidoctor and overriden it *incorrectly*.
But there are some cases where we really do want ot override it.
Specifically for logstash. So this allows us to turn it on when we want
it.

### CODE REVIEW

**Comment by nik9000** [comment]
@estolfo, could you have a look at this one?

@karenzone, I believe I need this one before converting logstash to asciidoctor.

File: resources/asciidoctor/lib/edit_me/extension.rb [hunk 1]
```diff
@@ -28,50 +28,73 @@ def process(document)
       url = url[0..-2] if url.end_with? '/'
       edit_urls << { toplevel: toplevel, url: url }
     end
-    edit_urls.reverse!
+    # Prefer the longest matching edit url
+    edit_urls = edit_urls.sort_by { |e| [-e[:toplevel].length, e[:toplevel]] }
```
Comment (estolfo) [inline resources/asciidoctor/lib/edit_me/extension.rb]: you could also use [max](https://ruby-doc.org/core-2.6.3/Enumerable.html#method-i-max), using the length as the comparable attribute in the supplied block.

File: resources/asciidoctor/lib/edit_me/extension.rb [hunk 2]
```diff
@@ -28,50 +28,73 @@ def process(document)
...
+        source_location.file
+      end
+    end
+  end
+
+  ##
+  # Extension to blocks that need an "edit me" link.
+  module WithEditLink
+    def title
+      url = edit_url
```
Comment (estolfo) [inline resources/asciidoctor/lib/edit_me/extension.rb]: Is there a reason you set the `url` variable here?
You can do `return super unless edit_url`
Then use the `edit_url` in the string as you've done in line 63.

File: resources/asciidoctor/lib/edit_me/extension.rb [hunk 3]
```diff
@@ -28,50 +28,73 @@ def process(document)
...
+    def edit_url
+      if @document.attributes['respect_edit_url_overrides']
+        url = @document.attributes['edit_url']
+        return false if url == ''
+        return url if url
+      end
+
+      edit_url_by_path
+    end
+
```
Comment (estolfo) [inline resources/asciidoctor/lib/edit_me/extension.rb]: The `if` block will return the last evaluated statement so you could also write this:
```ruby
def edit_url
  if @document.attributes['respect_edit_url_overrides']
    url = @document.attributes['edit_url']
    return false if url == ''
    url
  end || edit_url_by_path
end
```

File: resources/asciidoctor/lib/edit_me/extension.rb [hunk 4]
```diff
@@ -28,50 +28,73 @@ def process(document)
...
-        # the base_dir which we use when we build books from many repos.
-        document.source_location.file
-      end
-    else
-      def block.source_path
-        # source_location.path doesn't work for relative includes outside of
-        # the base_dir which we use when we build books from many repos.
-        source_location.file
-      end
+      url + path[entry[:toplevel].length..-1]
```
Comment (estolfo) [inline resources/asciidoctor/lib/edit_me/extension.rb]: I wouldn't use so many `return`s here and instead use `if-else` like this:

```ruby
if entry
  url = entry[:url]
  if url == '<disable>'
    false
  else
    url + path[entry[:toplevel].length..-1]
  end
else
  logger.warn#...etc
  false
end
```

**Comment by nik9000** [comment]
> Hi Nik, I made some suggestions for the code to be more idiomatic Ruby.

Thanks for reviewing! I'll do it!

File: resources/asciidoctor/lib/edit_me/extension.rb (hunk 1 above)
Comment (nik9000) [reply resources/asciidoctor/lib/edit_me/extension.rb]: I really want to sort the whole list so the longest are first so we check them first.

Comment (estolfo) [reply resources/asciidoctor/lib/edit_me/extension.rb]: ah, ok. I didn't see the big picture - you want a sorted list, not just the longest. Nevermind!

**Comment by nik9000** [comment]
@estolfo, I've pushed an update with the changes you requested. Please have another look when you get the chance!

**Comment by estolfo** [comment]
looks good!

**Comment by nik9000** [comment]
Thanks for reviewing @estolfo !

**Pull request merged**